round-based/src/mpc/mod.rs

```diff
@@ -0,0 +1,329 @@
+//! Party of MPC protocol
+//!
+//! [`MpcParty`] is party of MPC protocol, connected to network, ready to start carrying out the protocol.
+//!
+//! ```rust
+//! use round_based::{Incoming, Outgoing};
+//!
+//! # #[derive(round_based::ProtocolMsg)]
+//! # enum KeygenMsg {}
+//! # struct KeyShare;
+//! # struct Error;
+//! # type Result<T> = std::result::Result<T, Error>;
+//! # async fn doc() -> Result<()> {
+//! async fn keygen<M>(party: M, i: u16, n: u16) -> Result<KeyShare>
+//! where
+//!     M: round_based::Mpc<Msg = KeygenMsg>
+//! {
+//!     // ...
+//! # unimplemented!()
+//! }
+//! async fn connect() ->
+//!     impl futures::Stream<Item = Result<Incoming<KeygenMsg>>>
+//!         + futures::Sink<Outgoing<KeygenMsg>, Error = Error>
+//!         + Unpin
+//! {
+//!     // ...
+//! # round_based::_docs::fake_delivery()
+//! }
+//!
+//! let delivery = connect().await;
+//! let party = round_based::mpc::connected(delivery);
+//!
+//! # let (i, n) = (1, 3);
+//! let keyshare = keygen(party, i, n).await?;
+//! # Ok(()) }
+//! ```
+
+use crate::{
+    round::{RoundInfo, RoundStore},
+    Outgoing, PartyIndex,
+};
+
+pub mod party;
+
+#[doc(no_inline)]
+pub use self::party::{Halves, MpcParty};
+
+/// Abstracts functionalities needed for MPC protocol execution
+pub trait Mpc {
+    /// Protocol message
+    type Msg;
+
+    /// Returned in [`Self::finish_setup`]
+    type Exec: MpcExecution<Msg = Self::Msg, SendErr = Self::SendErr>;
+    /// Error indicating that sending a message has failed
+    type SendErr;
+
+    /// Registers a round
+    fn add_round<R>(&mut self, round: R) -> <Self::Exec as MpcExecution>::Round<R>
+    where
+        R: RoundStore,
+        Self::Msg: RoundMsg<R::Msg>;
+
+    /// Completes network setup
+    ///
+    /// Once this method is called, no more rounds can be added,
+    /// but the protocol can receive and send messages.
+    fn finish_setup(self) -> Self::Exec;
+}
+
+/// Abstracts functionalities needed for MPC protocol execution
+pub trait MpcExecution {
+    /// Witness that round was registered
+    ///
+    /// It is used to retrieve messages in [`MpcExecution::complete`].
+    type Round<R: RoundInfo>;
```

Comment:
I think this doc can be improved, but I don't know how. Currently it gives no guidance of what types can go there when implementing the `MpcExecution` trait, and after reading this MR I can't give this guidance either

Reply:
Expanded it a bit